doc/sphinx/proof-engine/ltac.rst

```diff
@@ -713,339 +962,563 @@ tactic
+
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros A B H.
+         match goal with
+         | H1 : _, H2 : _ |- _ => idtac "match " H1 H2; fail
+         | _ => idtac
+         end.
+
+   .. todo need examples for:
+      match_context_rule ::= [ {*, @match_hyp } |- @match_pattern ] => @ltac_expr
+      match_hyp ::= | @name := {? [ @match_pattern ] : } @match_pattern
+
+.. todo PR The following items (up to numgoals) are part of "value_tactic".  I'd like to make
+   this a subsection and explain that they all return values.  How do I get a 5th-level section title?
 
 Filling a term context
 ~~~~~~~~~~~~~~~~~~~~~~
 
 The following expression is not a tactic in the sense that it does not
 produce subgoals but generates a term to be used in tactic expressions:
 
-.. tacn:: context @ident [@ltac_expr]
+.. tacn:: context @ident [ @term ]
+
+   .. todo @term is actually @lconstr (term with holes), which is spliced away
+      Seems like it's worth restoring that distinction
 
-   :n:`@ident` must denote a context variable bound by a context pattern of a
-   match expression. This expression evaluates replaces the hole of the
-   value of :n:`@ident` by the value of :n:`@ltac_expr`.
+   Replaces the context variable :n:`@ident` bound by a `context` pattern of a
+   match expression (Described :ref:`here<match_term_context>`).
+   This expression evaluates replaces the hole of the
+   value of :n:`@ident` by the value of :n:`@term`.  See the example below.
+
+   :tacn:`context` is a :token:`value_tactic`.
 
    .. exn:: Not a context variable.
       :undocumented:
 
    .. exn:: Unbound context identifier @ident.
       :undocumented:
 
+   .. example:: Substituting a matched context
+
+      .. coqtop:: reset all
+
+         Goal True /\ True.
+         match goal with
+         | |- context G [True] => let x := context G [False] in idtac x
+         end.
+
 Generating fresh hypothesis names
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Tactics sometimes have to generate new names for hypothesis. Letting the
-system decide a name with the intro tactic is not so good since it is
-very awkward to retrieve the name the system gave. The following
+Tactics sometimes need to generate new names for hypothesis.  Letting |Coq|
+choose a name with the intro tactic is not so good since it is
+very awkward to retrieve that name. The following
 expression returns an identifier:
 
-.. tacn:: fresh {* @component}
+.. tacn:: fresh {* {| @string | @qualid } }
 
-   It evaluates to an identifier unbound in the goal. This fresh identifier
-   is obtained by concatenating the value of the :n:`@component`\ s (each of them
-   is, either a :n:`@qualid` which has to refer to a (unqualified) name, or
-   directly a name denoted by a :n:`@string`).
+   Evaluates to an identifier unbound in the goal. This fresh identifier
+   is formed by concatenating the value of each :n:`@qualid`,
+   which has to refer to an unqualified name, and :n:`string`.
```

Comment:
Do you understand what this means? Because I don't.
I was expecting the following to yield `bamHbam : True` but in fact it yields `barHbar : True`:

```coq
Module foo.
Ltac bar := ident:(bam).
End foo.

Goal True -> True.
let H := fresh foo.bar "H" foo.bar in intro H.
```

Reply:
Based on your example, I think it should say something like
```suggestion
   Evaluates to an identifier unbound in the goal. This fresh identifier
   is formed by concatenating the name of each :n:`@qualid`, after dropping any qualified components,
   and the value of each :n:`string`.
```

Reply:
If it just drops the qualifiers, we should consider changing the syntax to use `ident` instead of `qualid`.

Reply:
Also it should complain earlier about this:

```
let x := fresh X ";" in idtac x.

Invalid character ';' in identifier "X;".
```

Reply:
Both of these seem like things worth opening bug reports about.